src/riak_kv_token_session.erl

```diff
@@ -0,0 +1,467 @@
+    RT0 = os:system_time(millisecond),
+    case session_request(TokenID, NVal, RequestTO, TokenTO) of
+        {true, SessionRef} ->
+            {true, SessionRef};
+        _Error ->
+            timer:sleep(
+                rand:uniform(max(1, RequestTO div (Retry - Attempts)))
+            ),
+            session_request_retry(
+                TokenID,
+                NVal,
+                max(0, RequestTO + (RT0 - os:system_time(millisecond))),
+                TokenTO,
+                Retry,
+                Attempts + 1
+            )
+    end.
+
+-spec session_request(token_id(), 1|3|5|7, timeout_ms(), timeout_ms())
+        -> {true, session_ref()}|{false, none}|erpc_error().
+session_request(TokenID, NVal, RequestTimeout, TokenTimeout)
+        when NVal ==1; NVal == 3; NVal == 5; NVal == 7 ->
+    DocIdx = chash_key(TokenID),
+    PrimPartitions = riak_core_apl:get_primary_apl(DocIdx, NVal, riak_kv),
+    TargetNodeCount =
+        case NVal of
+            1 -> 1;
+            TNC when TNC > 1 -> TNC -2
+        end,
+    PrimNodes =
+        lists:uniq(
+            lists:map(fun({{_Idx, N}, primary}) -> N end, PrimPartitions)
+        ),
+    case length(PrimNodes) of
+        L when L >= TargetNodeCount ->
+            [Head|VerifyList] = lists:sublist(PrimNodes, TargetNodeCount),
+            case Head of
+                ThisNode when ThisNode == node() ->
+                    session_local_request(
+                        TokenID, VerifyList, RequestTimeout, TokenTimeout);
+                _ ->
+                    try
+                        erpc:call(
+                            Head,
+                            ?MODULE,
+                            session_local_request,
+                            [TokenID, VerifyList, RequestTimeout, TokenTimeout]
+                        )
+                    catch
+                        error:{erpc,noconnection}:_ ->
+                            ?LOG_WARNING(
+                                "Connection error"
+                                " accessing token_manager on ~w",
+                                [Head]),
+                            riak_kv_stat:update(token_session_unreachable),
+                            {false, none}
+                    end
+            end;
+        _ ->
+            ok = riak_kv_stat:update(token_session_preflist_short),
+            {false, none}
+    end.
+
+-spec session_use(session_ref(), atom(), list()) -> any().
+session_use(SessionReference, FuncName, Args) ->
+    {N, P, ID} = decode_session_reference(SessionReference),
+    case node() of
+        N ->
+            session_local_use(P, FuncName, Args, ID);
+        _ ->
+            safe_erpc(N, ?MODULE, session_local_use, [P, FuncName, Args, ID])
+    end.
+
+-spec session_release(session_ref()|none) -> ok|erpc_error().
+session_release(none) ->
+    ok;
+session_release(SessionReference) ->
+    {N, P, ID} = decode_session_reference(SessionReference),
+    case node() of
```

Comment:
Note that after restaring a node, the Pid `P` may have been gone, which is fine... but in very rare circumstances, it could also be the case that that node actually has this Pid, but that that process is running some totally different code. Very unlikely, but....

What happens then is that an arbitrary Process may be getting a gen_server:call from `sessio_local_use`.  Since that has timeout ìnfinity` and since it may be a process that just discards the message, you may have a deadlock there, waiting for every to use the session that no longer exists.